x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/plugin/TransportActionUtils.java

```diff
@@ -78,4 +78,56 @@ public static void executeRequestWithRetryAttempt(
             queryRunner.accept(onFailure);
         }
     }
+
+    public static void retryOnNodeWithMatchingVersion(
+        ClusterService clusterService,
+        QlVersionMismatchException exception,
+        Consumer<DiscoveryNode> retryRequest,
+        Consumer<Exception> onFailure,
+        Logger log
+    ) {
+        if (log.isDebugEnabled()) {
```

Comment:
(very minor note) IMHO this log does not belong here. This method is responsible of retrying the request, not of checking the version.